### PR TITLE
Make LRUQueryCache respect Accountable queries on eviction and consisten…

### PR DESCRIPTION
**Pull request by gtroitskiy**
…cy check

**Root cause**
`onQueryCache` increases `ramBytesUsed` for specified amount, that is being calculated with respect to query being `Accountable` or not.
Unfortunately, `onQueryEviction` does not the same. If some heavy accountable query has `ramBytesUsed()` greater than `QUERY_DEFAULT_RAM_BYTES_USED`, the delta `ghostBytes = ramBytesUsed() - QUERY_DEFAULT_RAM_BYTES_USED` remains in total `LRUQueryCache#ramBytesUsed` forever.

**Hit rate drops to 0**
since total sum of ghost bytes monotonously increases with each cached accountable query (>QUERY_DEFAULT_RAM_BYTES_USED), eventually it becomes greater than `maxRamBytesUsed`, and each newly cached query immediately evicts from the cache.

**Current behavior of LRUQueryCache for some real service in production [though not fully optimized]**
1. Service restarted.
2. Reached `maxRamBytesUsed`. Started eviction. From now on size of cached queries decreases to compensate increasing ghost bytes.
3. Total amount of ghost bytes is greater than `maxRamBytesUsed`. If any new query was cached, it evicted at the same time. Cache size is 0. Hit rate is 0. 
ram
size2
hitrate

**After fix**
image

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/search/LRUQueryCache.java [hunk 1]
```diff
@@ -385,7 +385,9 @@ public void clearQuery(Query query) {
 
   private void onEviction(Query singleton) {
     assert lock.isHeldByCurrentThread();
-    onQueryEviction(singleton, LINKED_HASHTABLE_RAM_BYTES_PER_ENTRY + QUERY_DEFAULT_RAM_BYTES_USED);
+    var ramBytesUsedByQuery = LINKED_HASHTABLE_RAM_BYTES_PER_ENTRY;
+    ramBytesUsedByQuery += singleton instanceof Accountable accountableQuery ? accountableQuery.ramBytesUsed() : QUERY_DEFAULT_RAM_BYTES_USED;
```
Comment (romseygeek) [inline lucene/core/src/java/org/apache/lucene/search/LRUQueryCache.java:389]: Can we pull this out into a separate `queryRamBytesUsed(Query q)` method, given that it's used twice?

**Review by romseygeek — CHANGES_REQUESTED** [review]
Thanks for finding and fixing this bug!  I left one small request for change but looks good to me otherwise.

**Comment by romseygeek** [comment]
Can you run `./gradlew tidy` at the root of the project to make sure the formatting is all correct?

**Comment by gtroitskiy** [comment]
Thanks for reviewing! I ran tidy and made some refactoring

**Review by romseygeek — COMMENTED** [review]
Looks great, thank you!  One more request: can you add an entry to `lucene/CHANGES.txt` in the 'Bug Fixes' section for Lucene 9.9.0?

**Review by romseygeek — APPROVED** [review]

**Comment by justinmarygopal** [comment]
I am also facing the same issue  after the upgrade of  elastic to 8.10 from 8.7.
Then we upgraded to  8.12, still we are seeing a similar behaviour . Any suggestions please?

**Comment by jaebongim** [comment]
@gtroitskiy @romseygeek 
Is the bug fixed on 8.12 Elasticseach?

**Pull request merged**